pkg/pool-weighted/test/foundry/ManagedPoolTokenLib.t.sol

```diff
@@ -0,0 +1,103 @@
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import { Test } from "forge-std/Test.sol";
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/test/TestToken.sol";
+
+import "../../contracts/WeightedMath.sol";
+import "../../contracts/lib/ValueCompression.sol";
+
+import "../../contracts/test/MockManagedPoolTokenLib.sol";
+import "../../contracts/managed/ManagedPoolTokenLib.sol";
+
+contract ManagedPoolTokenLibTest is Test {
+    uint256 private constant _START_DENORM_WEIGHT_OFFSET = 0;
+    uint256 private constant _END_DENORM_WEIGHT_OFFSET = _START_DENORM_WEIGHT_OFFSET + _DENORM_WEIGHT_WIDTH;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = _END_DENORM_WEIGHT_OFFSET + _DENORM_WEIGHT_WIDTH;
+
+    uint256 private constant _DENORM_WEIGHT_WIDTH = 64;
+    uint256 private constant _DECIMAL_DIFF_WIDTH = 5;
+
+    uint256 private constant _MAX_DENORM_WEIGHT = 1e22;
+
+    MockManagedPoolTokenLib private mock;
+
+    function setUp() external {
+        mock = new MockManagedPoolTokenLib();
+    }
+
+    function clearWordAtPosition(
+        bytes32 word,
+        uint256 offset,
+        uint256 bitLength
+    ) internal returns (bytes32 clearedWord) {
+        uint256 mask = (1 << bitLength) - 1;
+        clearedWord = bytes32(uint256(word) & ~(mask << offset));
+    }
+
+    function assertOtherStateUnchanged(
+        bytes32 oldPoolState,
+        bytes32 newPoolState,
+        uint256 offset,
+        uint256 bitLength
+    ) internal returns (bool) {
+        bytes32 clearedOldState = clearWordAtPosition(oldPoolState, offset, bitLength);
+        bytes32 clearedNewState = clearWordAtPosition(newPoolState, offset, bitLength);
+        assertEq(clearedOldState, clearedNewState);
+    }
+
+    function testScalingFactor(bytes32 tokenState, uint8 decimals) external {
+        ERC20 token = new TestToken("Test", "TEST", decimals);
+
+        if (decimals <= 18) {
+            uint256 expectedScalingFactor = FixedPoint.ONE * 10**(18 - decimals);
+
+            bytes32 newTokenState = mock.setTokenScalingFactor(tokenState, token);
+            uint256 tokenScalingFactor = mock.getTokenScalingFactor(newTokenState);
+            assertEq(tokenScalingFactor, expectedScalingFactor);
+        } else {
+            vm.expectRevert("BAL#001"); // SUB_OVERFLOW
+            mock.setTokenScalingFactor(tokenState, token);
+        }
+    }
+
+    function testWeightChangeFields(
+        bytes32 tokenState,
+        uint64 normalizedStartWeight,
+        uint64 normalizedEndWeight
+    ) public {
```

Comment:
```suggestion
    ) public {
        vm.assume(normalizedStartWeight < FixedPoint.ONE);
        vm.assume(normalizedEndWeight < FixedPoint.ONE);
```

We almost get this for free since fp(1) is 60 bits, but it doesn't hurt to reject some.